File: ml_test_contest_04_07_2023/ml_test_contest/_3.cpp
#include <vector>
#include <fstream>
// #include <iostream>
using namespace std;

bool is_target(int a_i, int k) {
    return 1 <= a_i && a_i <= k;
}

void reset_dict(vector<int>& d) {
    for (size_t i = 0; i < d.size(); i++)
    {
        d[i] = 0;
    }
}

bool is_all_found(vector<int>& d, int k,int& current_found) {
    if (current_found<k)
    {
        return false;
    }
    
    current_found =0;
    for (size_t i = 0; i < d.size(); i++)
    {
        if (d[i]==1)
        {
            current_found +=1;
        }
        
    }
    if (current_found<k)
    {
        return false;
    }
    return true;
}


void search_next_seq(vector<int>& a_vec, int& start_pos, long long& min_s, int k, int n, vector<int>& have_y_seen_the_k) {
    long long current_s = 0;
    // |1| 2 1 4 2 1 |3| 7 8 9 11 1 2 3
    // 1 9 2 4 2 2 2 3 1 8 2 10 9
    int current_found = 0;
    for (int i = start_pos; i < n; ++i) {
        int a_i = a_vec[i];
        if (is_target(a_i, k)) {
            have_y_seen_the_k[a_i-1] = 1;
            current_found += 1;
        } else if (!is_target(a_i, k) && current_s == 0) {
            start_pos += 1;
            continue;
        }
        current_s += a_i;
        
        if (is_all_found(have_y_seen_the_k, k,current_found)) {
            min_s = min(min_s, current_s);
            reset_dict(have_y_seen_the_k);
            current_found=0;
            current_s = 0;
            for (int j = i; j > start_pos-1;--j)
            {
                // 6 2 3 1 2 3
                // 1 2 4 3 4 1 5 6
                // 1-4
                // 1 9 1 3 2 |4| 2 2 5 4 9 2 3 1 8 2 10 9
                // 1 9 1 3 2 4 2 2 5 4
                // 1 9 1 1 3 2 9 3
                
                int a_j = a_vec[j];
                if (is_target(a_j, k)) {
                    have_y_seen_the_k[a_j-1] = 1;
                    current_found += 1;
                }
                if (is_all_found(have_y_seen_the_k, k,current_found))
                {   
                    current_s += a_j;
                    min_s = min(min_s,current_s);
                    reset_dict(have_y_seen_the_k);
                    start_pos = j+1;
                    return;
                }
                current_s += a_j;                
            }
        }
    }
    start_pos = n;
}

int main() {
    ifstream ifstream("input.txt");
    ofstream ofstream("output.txt");
    int n, k;
    ifstream >> n >> k;
    if(k==1){
        ofstream << 1;
            ifstream.close();
            ofstream.close();
            return EXIT_SUCCESS; 
    }
    vector<int> a_vec(n);
    for (int i = 0; i < n; i++) {
        ifstream >> a_vec[i];
    }
    long long min_s = 125000750001;
    int start_pos = 0;
    vector<int> have_y_seen_the_k(k);
    while(true){
        search_next_seq(a_vec,start_pos,min_s,k,n, have_y_seen_the_k);
        if (start_pos > n-k){
            break;
        }
    }
    ofstream << min_s;
    ifstream.close();
    ofstream.close();
    return EXIT_SUCCESS; 
}

File: ml_contest_07_07_2023/_3.py
import sys
import numpy as np

def is_target(a_i: np.intc, k: int) -> bool:
    return 1 <= a_i <= k

def reset_dict(d: np.array) -> None:
    d.fill(0)
    # print(1)
    # for i in range(len(d)):
    #     d[i] = np.intc

def is_all_found(d: np.array, k: int) -> bool:
    # for el in d.values():
    #     s_ += el
    return np.sum(d) == k

def search_next_seq(a_vec:np.array, start_pos:int, min_s:int, k:int, n:int, have_y_seen_the_k: np.array):
    current_s = 0
    for i in range(start_pos, n):
        a_i = a_vec[i]
        if is_target(a_i, k):
            have_y_seen_the_k[a_i - 1] = 1
        elif (not is_target(a_i, k)) and current_s == 0:
            start_pos += 1
            continue
        current_s += a_i

        # else:
        #     current_s = 0
        #     reset_dict(have_y_seen_the_k)
        #     start_pos = i + 1
        if is_all_found(have_y_seen_the_k, k):
            min_s = min(min_s, current_s)
            reset_dict(have_y_seen_the_k)
            start_pos += 1
            return start_pos, min_s
    return n, min_s


if __name__ == '__main__':
    ifstream = open('input.txt', 'r')
    ofstream = open('output.txt', 'w')
    n, k = tuple([int(el) for el in ifstream.readline().replace('\n', '').split(' ')])
    a_vec = np.array([int(el) for el in ifstream.readline().replace('\n', '').split(' ')], dtype=np.intc)
    # should buy a_i if a_i in 1...k
    # несколько статуэток одного вида можно покупать.
    # задача: покупка статуэток непрерывным куском l...r
    # 1 2 4 5 3
    # хочет купить a_i in 1...3
    # купит все 1 2 4 5 3
    # 1 2 2 3 3 1
    # хочет купить a_i in 1...3
    # купит 1 2 2 3
    # цена статуэтки == a_i
    # найти минимальную по стоимости последовательность {a_j} подряд идущих a_i
    # таких, что  среди a_j есть все 1...k

    # len(a_vec) == n

    # 1. найти очередную последовательность, которая содержит все 1...k
    # 2. посчитать стоимость такой последовательности
    # 3. повторить п. 1-2 выбрать среди всех последовательностей минимальную стоимость
    # если нет подряд идущих то выдать хоть что-то

    # limitations:
    # n >= k

    min_s = int((5 * 10 ** 5 + 1) * (5 * 10 ** 5 + 2) / 2)
    # have_y_seen_the_k = {i: 0 for i in range(1, k+1)}
    have_y_seen_the_k = np.zeros(shape=(k,), dtype=np.intc)
    # reset_dict(have_y_seen_the_k)
    start_pos = 0
    while True:
        start_pos, min_s = search_next_seq(a_vec, start_pos, min_s, k, n, have_y_seen_the_k)
        if start_pos > n - k:
            break

    ofstream.write(str(min_s))
    ifstream.close()
    ofstream.close()
    sys.exit(0)